osf/models/preprint.py

```diff
@@ -0,0 +1,920 @@
+
+        for group_name in self.groups.keys():
+            if self.get_group(group_name).user_set.filter(id=old.id).exists():
+                self.get_group(group_name).user_set.remove(old)
+                self.get_group(group_name).user_set.add(new)
+        return res
+
+    # Overrides ContributorMixin to run update_search
+    def move_contributor(self, contributor, auth, index, save=False):
+        super(Preprint, self).move_contributor(contributor, auth, index, save)
+        self.update_search()
+
+    # Overrides ContributorMixin since this query is constructed differently
+    def _get_admin_contributors_query(self, users):
+        return PreprintContributor.objects.select_related('user').filter(
+            preprint=self,
+            user__in=users,
+            user__is_active=True,
+            user__groups=(self.get_group('admin').id))
+
+    # Overrides ContributorMixin to run update_search
+    def set_visible(self, user, visible, log=True, auth=None, save=False):
+        super(Preprint, self).set_visible(user, visible, log=log, auth=auth, save=save)
+        self.update_search()
+
+    def on_update(self, first_save, saved_fields):
+        User = apps.get_model('osf.OSFUser')
+        request, user_id = get_request_and_user_id()
+        request_headers = {}
+        if not isinstance(request, DummyRequest):
+            request_headers = {
+                k: v
+                for k, v in get_headers_from_request(request).items()
+                if isinstance(v, basestring)
+            }
+
+        user = User.load(user_id)
+        if user and self.check_spam(user, saved_fields, request_headers):
+            # Specifically call the super class save method to avoid recursion into model save method.
+            super(Preprint, self).save()
+
+    @classmethod
+    def bulk_update_search(cls, preprints, index=None):
+        from website import search
+        try:
+            serialize = functools.partial(search.search.update_preprint, index=index, bulk=True, async=False)
+            search.search.bulk_update_nodes(serialize, preprints, index=index)
+        except search.exceptions.SearchUnavailableError as e:
+            logger.exception(e)
+            log_exception()
+
+    def update_search(self):
+        from website import search
+        try:
+            search.search.update_preprint(self, bulk=False, async=True)
+        except search.exceptions.SearchUnavailableError as e:
+            logger.exception(e)
+            log_exception()
+
+    def create_root_folder(self):
+        if self.root_folder:
+            return self.root_folder
+
+        # Note: The "root" node will always be "named" empty string
+        root_folder = OsfStorageFolder(name='', target=self, is_root=True)
+        root_folder.save()
+
+        self.root_folder = root_folder
+        self.save()
+        return root_folder
+
+    def serialize_waterbutler_settings(self):
```

Comment:
Since preprints don't have addons, there are a few osfstorage methods we need to copy here, `serialize_waterbutler_settings`, `serialize_waterbutler_credentials`, `create_waterbutler_log`.